pyxdsm/XDSM.py

```diff
@@ -19,744 +23,536 @@
-            controls how many items in the tuple/list will be displayed per line.
-            If None, the label will be printed one item per line if given as a tuple or list,
-            otherwise the string will be printed on a single line.
-
-        spec_name : str
-            The spec name used for the spec file.
-
+        use_sfmath : bool
+            Whether to use the sfmath latex package
+        optional_latex_packages : str or list of strings
+            Additional latex packages for PDF/TEX generation
+        auto_fade : dict
+            Auto-fade configuration with keys: inputs, outputs, connections, processes
         """
-        if spec_name is None:
-            spec_name = node_name
-
-        sys = System(node_name, style, label, stack, faded, label_width, spec_name)
-        self.systems.append(sys)
+        # Only process if these aren't already in data (from deserialization)
+        if "optional_packages" not in data:
+            # Process optional packages
+            packages = []
+            if optional_latex_packages is not None:
+                if isinstance(optional_latex_packages, str):
+                    packages = [optional_latex_packages]
+                elif isinstance(optional_latex_packages, list):
+                    packages = optional_latex_packages
+                else:
+                    raise ValueError("optional_latex_packages must be a string or list of strings")
+            data["optional_packages"] = packages
+
+        if "auto_fade" not in data:
+            # Process auto_fade
+            fade_config = AutoFadeConfig()
+            if auto_fade is not None:
+                fade_config = AutoFadeConfig(**auto_fade)
+            data["auto_fade"] = fade_config
+
+        if "use_sfmath" not in data:
+            data["use_sfmath"] = use_sfmath
+
+        super().__init__(**data)
+
+    @model_validator(mode="before")
+    @classmethod
+    def _set_defaults_for_missing_fields(cls, data):
+        """Ensure missing or null collection fields get empty defaults."""
+        if not isinstance(data, dict):
+            return data
+
+        # Set empty defaults for missing or null collection fields
+        if "inputs" not in data or data.get("inputs") is None:
+            data["inputs"] = {}
+        if "outputs" not in data or data.get("outputs") is None:
+            data["outputs"] = {}
+        if "systems" not in data or data.get("systems") is None:
+            data["systems"] = []
+        if "connections" not in data or data.get("connections") is None:
+            data["connections"] = []
+        if "processes" not in data or data.get("processes") is None:
+            data["processes"] = []
+
+        return data
+
+    @model_validator(mode="after")
+    def _validate_unique_system_names(self):
+        """Ensure all system names are unique."""
+        names = [sys.node_name for sys in self.systems]
+        duplicates = [n for n in names if names.count(n) > 1]
```

Comment:
A slightly simpler option:
```
if len(set(names)) < len(names):
    ...
```

Reply:
True, but this way the validation error tells which system names are duplicated.